### PR TITLE
ZQS-1336: Inverse of Block Encoding

### PR DESCRIPTION
**Pull request by olgOk**
## Description

- Implementation of the QSP in qiskit (qsp.py).

-  Added the estimation of the parameters for quantum DE solver: number of grid points, matrix condition number, degree of approximating polynomial.

-  Construction of the inverse of block encoding (SEL_inv) using the QSP circuit construction.

## Please verify that you have completed the following steps

- [x] I have self-reviewed my code.
- [x] I have included test cases validating introduced feature/fix.
- [x] I have updated documentation.

### CODE REVIEW

**Review by AthenaCaesura — APPROVED** [review]
Just a few minor comments. 🙂

File: src/benchq/algorithms/lde_solver.py [hunk 1]
```diff
@@ -78,9 +159,67 @@ def control_prep(k: int, beta: float) -> Tuple[Circuit, Circuit]:
...
+    # number of grid points
+    k = get_num_of_grid_points(matrix_norm, epsilon, beta)
+    kappa = get_kappa(matrix_norm, time)
+    num_phis = get_degree(kappa, epsilon)
+    # generate arbitrary phases
+    phi_seq = np.linspace(-np.pi, np.pi, num_phis)
+
+    prep, prep_prime = control_prep(k=k, beta=beta)
+    prep_prime_dag = prep_prime.inverse()
+    grid_qubits = int((np.ceil(np.log2(k)))) + 1
```
[the remark this answers is not in the document]
Comment (olgOk) [reply src/benchq/algorithms/lde_solver.py:204]: np.ceil returns 'float' datatype but I the number of qubits to be 'int' type.

File: src/benchq/algorithms/qsp.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+https://github.com/qsppack/RACBEM/blob/master/racbem.py
+
+The work is based on the research paper by Yulong Dong and Lin Lin.
+"Random circuit block-encoded matrix and a proposal of quantum
+LINPACK benchmark."
+https://arxiv.org/abs/2006.04010
+
+Authors:
+    Yulong Dong     dongyl (at) berkeley (dot) edu
+    Lin Lin         linlin (at) math (dot) berkeley (dot) edu
```
[the remark this answers is not in the document]
Comment (olgOk) [reply src/benchq/algorithms/qsp.py:47]: I took this as it was in the original repo, but I will do something about it if needed.

File: src/benchq/algorithms/lde_solver.py [hunk 3]
```diff
@@ -1,7 +1,86 @@
...
+    For the matrix inversion problem, the function to be approximated is 1/x.
+
+    The formula is given by the equations (35)-(36) as described in the paper
+    by Y. Dong, X. Meng, K. B. Whaley, and L. Lin.
+    "Efficient Phase Factor Evaluation in Quantum Signal Processing".
+    https://arxiv.org/pdf/2002.11649.pdf
+
+    Note that in this implementation the degree of polynomial is devided
+    by the factor of 3 to reflect a result of a possible improvements
+    by Remez algorithm.
```
[the remark this answers is not in the document]
Comment (olgOk) [reply src/benchq/algorithms/lde_solver.py:39]: it appears to be in French... I will include Wikipedia's page

File: src/benchq/algorithms/lde_solver.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (AthenaCaesura) [reply src/benchq/algorithms/lde_solver.py:204]: Then you can just use the builtin `ceil` function.

**Review by AthenaCaesura — CHANGES_REQUESTED** [review]

File: src/benchq/algorithms/qsp.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (AthenaCaesura) [reply src/benchq/algorithms/qsp.py]: Nah, the authors probably did that to prevent bots from getting thier emails I guess. No need.

**Review by AthenaCaesura — APPROVED** [review]

File: src/benchq/algorithms/lde_solver.py [hunk 4]
```diff
@@ -84,3 +164,54 @@ def control_prep(k: int, beta: float) -> Tuple[Circuit, Circuit]:
...
+        be_matrix (Circuit): the block encoding of a matrix.
+        matrix_norm (float): the norm of the matrix that is
+            to be block encoded.
+        time (float): the time interval one seeks solution
+            for a differntial equation.
+        beta (float): a bounding parameter of the matrix.
+        epsilon (float): an accuracy for the polynomial approximation.
+
+        Returns:
+            sel_inverse (Circuit): an instance of oqquestra quantum circuit
```
Comment (max-radin) [inline src/benchq/algorithms/lde_solver.py:184]:
```suggestion
            sel_inverse (Circuit): an Orquestra quantum circuit
```

**Review by max-radin — COMMENTED** [review]
Looks great! I left a couple minor comments about the docstrings. Also, ideally we would have a unit test to check that `inverse_blockencoding` works correctly. Is there some choice of `be_matrix`, `matrix_norm`, etc. for which it's straightforward to know what the resulting circuit should be? Or know some property of the circuit?

File: src/benchq/algorithms/lde_solver.py [hunk 5]
```diff
@@ -1,8 +1,88 @@
...
+    Returns:
+        kappa (float): the condition number of Xi.
+    """
+    return np.exp(2 * time_interval * matrix_norm)
+
+
+def get_degree(kappa: float, epsilon: float) -> int:
+    """Calculates an estimation of the degree of an approximating polynomial
+    such that the polynomial is epsilon-close to the desired function.
+    For the matrix inversion problem, the function to be approximated is 1/x.
```
[the remark this answers is not in the document]
Comment (olgOk) [reply src/benchq/algorithms/lde_solver.py:35]: Yes, it is only for `1/x`. The formula would be different if approximating other function.

[the remark this answers is not in the document]
Comment (olgOk) [reply src/benchq/algorithms/lde_solver.py:35]: I will adjust this docstring with the next PR. Thank you for pointing this out

**Pull request merged**